Run upgrade twice fixes #5377

albertlast opened the pull request:
Thanks to @Sesquipedalian we could now create upgrade file which do a upgrade check in a easy fashion

One additional thing,
i removed the ' from "// PHP won't suppress" with itention,
because it's create a syntax highilight issue --> syntax highlight > right writing

atm is the code not tested

issue #5377

Sesquipedalian committed:
Adds ability to skip DB change substeps in the upgrader
If a PHP code block within one of the upgrade*.sql files sets $upcontext['skipStep'] to true, then the rest of the substeps within the current main step of the database changes will be skipped. Once the upgrader reaches the next main step (meaning, a line formatted like `--- Adding new settings...`), $upcontext['skipStep'] will be set back to false, and the upgrader will continue normally. This is especially useful for conducting sanity checks before applying changes.

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

albertlast committed:
Added upgrade check to all new stuff

Signed-off-by: albertlast albertlast@hotmail.de

albertlast committed:
Do some stuff in the old env

Signed-off-by: albertlast albertlast@hotmail.de

albertlast commented:
well the second commit from my side did some stuff where i'm not sure which route i want to go...
when we insert values and ther already values with this name.
Want we delete them or keep them untouched?
atm i delete/override them.

Oldiesmann commented:
We need to leave them alone. We especially don't want to change things in the settings table unless absolutely necessary as that will only annoy users.

Sesquipedalian committed:
Fixes bugs with skipping substeps in database changes

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

albertlast committed:
Merge remote-tracking branch 'sesqui/upgrader_skipstep' into RunUpgradeTwice

albertlast committed:
add more checks

Signed-off-by: albertlast albertlast@hotmail.de

albertlast committed:
removed harmfull statements

Signed-off-by: albertlast albertlast@hotmail.de

albertlast commented:
thanks @Sesquipedalian i was able to change the upgrader in a way where no sql error appears in the pg log while running upgrader multiple times.

i used three major approach to checks
for column exists
```
---# upgrade check
---{
$table_columns = $smcFunc['db_list_columns']('{db_prefix}log_group_requests');
$upcontext['skipStep'] = in_array('act_reason', $table_columns);
---}
---#
```

for table exists
```
---# upgrade check
---{
$table_columns = $smcFunc['db_list_columns']('{db_prefix}user_drafts');
$upcontext['skipStep'] = count($table_columns) > 0 ? true : false;
---}
---#
```

for column typ check
```
---# upgrade check
---{
$column_info = upgradeGetColumnInfo('{db_prefix}member_logins', 'ip');
if (stripos($column_info['type'], 'inet') !== false)
	$upcontext['skipStep'] = true;
---}
---#
```

Sesquipedalian commented:
Excellent. Would you mind looking over the MySQL file as well?

albertlast committed:
transfer changes to mysql

Signed-off-by: albertlast albertlast@hotmail.de

albertlast commented:
added mysql too
mysql got no error log of sql statements so i only transfer the changes from pg to mysql

on pg side i need a test if the upgrade work from 2.0.x to 2.1.
the upgrade to upgrade is already tested

on mysql side both tests are open atm.

Sesquipedalian commented:
Thanks, @albertlast. Just to clarify, what does "both tests are open" mean?

albertlast commented:
mysql 2.0 to 2.1
mysql 2.1 to 2.1

Sesquipedalian commented:
Okay. But does saying that those tests are "open" mean that they have not been tested yet?

albertlast commented:
we need someone how test this,
yes.

Sesquipedalian commented:
When upgrading MySQL, using `$smcFunc['db_list_columns']()` to test whether a table exists causes the upgrade to halt. For example, the following code halts the process during a 2.0 → 2.1 upgrade because the member_logins table does not exist when it is run:
```
---# upgrade check
---{
$table_columns = $smcFunc['db_list_columns']('{db_prefix}member_logins');
$upcontext['skipStep'] = count($table_columns) > 0 ? true : false;
---}
---#
```

Sesquipedalian commented:
Since we can use `CREATE TABLE IF NOT EXISTS` in both MySQL and PostgreSQL, maybe we don't need to bother with a test before doing that. Alternatively, `$smcFunc['db_list_tables']()` from DbExtra-*.php might be appropriate here.

Sesquipedalian commented:
By the way, it is okay to do a check in the middle of a step. For example, in the following, the first and second sub-steps would be performed, but the third and fourth sub-steps would not:
```
/******************************************************************************/
--- Foo...
/******************************************************************************/
---# first sub-step...
< SQL code goes here >
---#

---# second sub-step...
---{
< php code goes here >
---}
---#

---# upgrade check
---{
$bar = false;
$upcontext['skipStep'] = $bar;
---}
---#

---# third sub-step...
< SQL code goes here >
---#

---# fourth sub-step...
---{
< php code goes here >
---}
---#

```

albertlast commented:
I use the table check as sign/ indicator if the step needs to be run or not.


________________________________
Von: Jon Stovell <notifications@github.com>
Gesendet: Samstag, 2. Februar 2019 01:12
An: SimpleMachines/SMF2.1
Cc: albertlast; Mention
Betreff: Re: [SimpleMachines/SMF2.1] Run upgrade twice fixes #5377 (#5383)


Since we can use CREATE TABLE IF NOT EXISTS in both MySQL and PostgreSQL, maybe we don't need to bother with a test before doing that. Alternatively, $smcFunc['db_list_tables']() from DbExtra-*.php might be appropriate here.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/SimpleMachines/SMF2.1/pull/5383#issuecomment-459910964>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABs0eiExajm2w7N0qhh384lYEK2aX1BDks5vJNfXgaJpZM4acYnV>.

albertlast commented:
I use this already on one place,
But I don't like it.
Because from maintenance pov it's harder to find/see.


________________________________
Von: Jon Stovell <notifications@github.com>
Gesendet: Samstag, 2. Februar 2019 01:21
An: SimpleMachines/SMF2.1
Cc: albertlast; Mention
Betreff: Re: [SimpleMachines/SMF2.1] Run upgrade twice fixes #5377 (#5383)


By the way, it is okay to do a check in the middle of a step. For example, in the following, the first and second sub-steps would be performed, but the third and fourth sub-steps would not:

/******************************************************************************/
--- Foo...
/******************************************************************************/
---# first sub-step...
< SQL code goes here >
---#

---# second sub-step...
---{
< php code goes here >
---}
---#

---# upgrade check
---{
$bar = false;
$upcontext['skipStep'] = $bar;
---}
---#

---# third sub-step...
< SQL code goes here >
---#

---# fourth sub-step...
---{
< php code goes here >
---}
---#



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/SimpleMachines/SMF2.1/pull/5383#issuecomment-459912485>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABs0epm4lxQY3tnAIv-sFHou4WJwQzQqks5vJNn_gaJpZM4acYnV>.

Sesquipedalian commented:
> I use the table check as sign/ indicator if the step needs to be run or not.

Well, we either need to use a different function to do the test when the table may not exist yet, or else we need to `CREATE TABLE IF EXISTS` before using this function.

Sesquipedalian commented:
I think it would be best to use an `$upcontext['skipStep'] = true;` test only in places where it is absolutely necessary in order to avoid errors. In places where the existing logic can already handle the different possibilities, such as `CREATE TABLE IF EXISTS` statements, we should just leave it alone.

Sesquipedalian commented on other/upgrade_2-1_mysql.sql:
```diff
@@ -30,13 +30,29 @@ WHERE YEAR(birthdate) < 1004;
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}log_spider_stats');
+$upcontext['skipStep'] = !in_array('state_date', $table_columns);
```
1. Why test this one but none of the others?
2. Do we need to bother testing this anyway? I don't think altering the default value of a column to the same default it already has will cause any errors, so we can probably just leave it alone.

Sesquipedalian commented on other/upgrade_2-1_mysql.sql:
```diff
@@ -1190,6 +1268,13 @@ ALTER TABLE `{$db_prefix}members`
 /******************************************************************************/
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}user_drafts');
```
This will fail and cause the upgrader to halt if the table doesn't exist. Also, I don't think we need to test this anyway.

Sesquipedalian commented on other/upgrade_2-1_mysql.sql:
```diff
@@ -1291,6 +1383,13 @@ ADD COLUMN likes SMALLINT(5) UNSIGNED NOT NULL DEFAULT '0';
 /******************************************************************************/
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}mentions');
```
This will fail and cause the upgrader to halt if the table doesn't exist. I don't think we need to test this.

Sesquipedalian commented on other/upgrade_2-1_mysql.sql:
```diff
@@ -1462,6 +1561,13 @@ $smcFunc['db_free_result']($file_check);
 /******************************************************************************/
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}qanda');
```
This will fail and cause the upgrader to halt if the table doesn't exist. I don't think we need to test this.

Sesquipedalian commented on other/upgrade_2-1_mysql.sql:
```diff
@@ -1607,6 +1713,13 @@ $request = upgrade_query("
 /******************************************************************************/
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}pm_recipients');
```
I don't think we should do this test. The "Moving label info to new tables and updating rules" sub-step contains lots of different things that it needs to check and manipulate—much more than this simple test can account for.

Sesquipedalian commented on other/upgrade_2-1_postgresql.sql:
```diff
@@ -212,13 +212,29 @@ WHERE EXTRACT(YEAR FROM birthdate) < 1004;
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}log_spider_stats');
+$upcontext['skipStep'] = !in_array('state_date', $table_columns);
```
1. Why test this one but none of the others?
2. Do we need to bother testing this anyway? I don't think altering the default value of a column to the same default it already has will cause any errors, so we can probably just leave it alone.

Sesquipedalian commented on other/upgrade_2-1_postgresql.sql:
```diff
@@ -212,13 +212,29 @@ WHERE EXTRACT(YEAR FROM birthdate) < 1004;
 /******************************************************************************/
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}member_logins');
```
This will fail and cause the upgrader to halt if the table doesn't exist. I don't think we need to test this anyway.

Sesquipedalian commented on other/upgrade_2-1_postgresql.sql:
```diff
@@ -1383,6 +1454,13 @@ ALTER TABLE {$db_prefix}members
 /******************************************************************************/
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}user_drafts');
```
This will fail and cause the upgrader to halt if the table doesn't exist. I don't think we need to test this.

Sesquipedalian commented on other/upgrade_2-1_postgresql.sql:
```diff
@@ -1419,20 +1497,18 @@ if (@$modSettings['smfVersion'] < '2.1')
-				VALUES
-					" . $insert);
-		}
+		$smcFunc['db_insert']('replace',
```
Nice.

Sesquipedalian commented on other/upgrade_2-1_postgresql.sql:
```diff
@@ -1443,32 +1519,40 @@ if (@$modSettings['smfVersion'] < '2.1')
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}messages');
+$upcontext['skipStep'] = in_array('likes', $table_columns);
```
What if the likes column exists in the message table, but the user_likes table doesn't? This test should be moved to a position immediately before the "Adding count to the messages table" sub-step.

Sesquipedalian commented on other/upgrade_2-1_postgresql.sql:
```diff
@@ -1490,6 +1574,13 @@ ADD COLUMN likes smallint NOT NULL default '0';
 /******************************************************************************/
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}mentions');
```
This will fail and cause the upgrader to halt if the table doesn't exist. Instead, after the `CREATE TABLE`, maybe we should test whether the indexes already exist or not.

albertlast replied:
Like i said,
the upgrader is tested on pg.
so db_list_columns works nice here to test the existing of table

Sesquipedalian commented on other/upgrade_2-1_postgresql.sql:
```diff
@@ -1685,6 +1776,13 @@ $smcFunc['db_free_result']($file_check);
 /******************************************************************************/
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}qanda');
```
This will fail and cause the upgrader to halt if the table doesn't exist.

Sesquipedalian commented on other/upgrade_2-1_postgresql.sql:
```diff
@@ -1814,33 +1910,37 @@ $request = upgrade_query("
 /******************************************************************************/
+---# upgrade check
+---{
+$table_columns = $smcFunc['db_list_columns']('{db_prefix}pm_recipients');
```
I don't think we should do this test. The "Moving label info to new tables and updating rules" sub-step contains lots of different things that it needs to check and manipulate—much more than this simple test can account for.

albertlast committed:
MySQL fix smf_db_list_columns

use the sql standard way to retrieve the information,
changed the format of the select to the show fields

same like pg and works even when the table didn't exists (so the pg and mysql api work in the same way)

Signed-off-by: albertlast albertlast@hotmail.de

albertlast committed:
correct stat_date (not state_date)

Signed-off-by: albertlast albertlast@hotmail.de

albertlast committed:
Merge branch 'MysqlDbListCol' into RunUpgradeTwice

albertlast commented:
> 
> 
> I think it would be best to use an `$upcontext['skipStep'] = true;` test only in places where it is absolutely necessary in order to avoid errors. In places where the existing logic can already handle the different possibilities, such as `CREATE TABLE IF EXISTS` statements, we should just leave it alone.

yeah i see the point,
so what you define as necessary?
At the point I do all the places where sql error occured,
maybe switch to places only where upgrader stop/break?

another point would instead of checking "something" from the step,
i could check if the last changed is applied.

Sesquipedalian commented:
I'd define necessary as meeting at least one of these two conditions:

1. If an error while running the step causes the upgrade to halt or display an error message to the user.
2. Running the step as second time messes up the data and/or the database structure.

So, for SQL errors that we can recover from gracefully, there is no need for a check.

Sesquipedalian commented:
Checking the last sub-step is probably a good default, but the best choice will need to be decided on a case-by-case basis.

albertlast committed:
a more passive way

Signed-off-by: albertlast albertlast@hotmail.de

albertlast committed:
remove newline

Signed-off-by: albertlast albertlast@hotmail.de

albertlast committed:
clean mysql and changed pg

Signed-off-by: albertlast albertlast@hotmail.de

albertlast committed:
added board_permissions_view check

Signed-off-by: albertlast albertlast@hotmail.de

albertlast commented:
@Sesquipedalian changes are done

albertlast committed:
Fix ban items upgrade

Signed-off-by: albertlast albertlast@hotmail.de

Sesquipedalian commented:
I'm not done testing yet, but I noticed that smf_db_list_columns() needs to include `table_schema` in the WHERE clause, just in case the same column name appears in more than one MySQL database.
```
function smf_db_list_columns($table_name, $detail = false, $parameters = array())
{
	global $smcFunc, $db_prefix, $db_name;

	$table_name = str_replace('{db_prefix}', $db_prefix, $table_name);

	$result = $smcFunc['db_query']('', '
		SELECT column_name "Field", COLUMN_TYPE "Type", is_nullable "Null", COLUMN_KEY "Key" , column_default "Default", extra "Extra"
		FROM information_schema.columns
		WHERE table_name = {string:table_name}
			AND table_schema = {string:db_name}
		ORDER BY ordinal_position',
		array(
			'table_name' => $table_name,
			'db_name' => $db_name,
		)
	);
...
```

Sesquipedalian commented:
I can confirm that once the change described in my 
[previous commnent](https://github.com/SimpleMachines/SMF2.1/pull/5383#issuecomment-460097072) is made, running and re-running the upgrader works nicely for MySQL.

albertlast commented:
Will you create a pr with the schema name fix?
Because I got surprised that you merge this without adding this.

Sesquipedalian commented:
I did add it. I was just about to post a comment here letting you know that, but you beat me to it. 🙂

Sesquipedalian commented:
https://github.com/SimpleMachines/SMF2.1/commit/5b563e1e136a85a3b613cdb36bcf2143aa0e6e7a

Sesquipedalian commented:
I also decided to rename `$upcontext['skipStep']` to `$upcontext['skip_db_substeps']` in the merged version, just to help prevent confusion for developers reading the code in the future.
